File: sprint6/By languages/C++/lesson8code1.cpp
#include <iostream>
#include <queue>
#include <stack>
#include <vector>
using namespace std;

const int V = 100; // число вершин графа

vector<int> adjList[V];
string color[V];
int previous[V];
int distance[V];

void BFS(int s) {
    queue<int> planned;
    planned.push(s);
    color[s] = "gray";
    distance[s] = 0;
    while (!planned.empty()) {
        int u = planned.front();
        planned.pop();
        for (int i = 0; i < adjList[u].size(); i++) {
            int v = adjList[u][i];
            if (color[v] == "") {
                distance[v] = distance[u] + 1;
                previous[v] = u;
                color[v] = "gray";
                planned.push(v);
            }
        }
        color[u] = "black";
    }
}

stack<int> ShortestPath(int v) {
    stack<int> path;
    int current_vertex = v;
    while (current_vertex != -1) {
        path.push(current_vertex);
        current_vertex = previous[current_vertex];
    }
    return path;
}

int main() {
    // add edges to adjList
    BFS(0);
    stack<int> path = ShortestPath(4);
    while (!path.empty()) {
        cout << path.top() << " ";
        path.pop();
    }
    cout << endl;
    return 0;
}

File: sprint6/By languages/JS/lesson8code1.js
// Длины массивов равны числу вершин|V|.
const color = new Array(V);
const previous = new Array(V);
const distance = new Array(V);

function BFS(s) {
    // Создадим очередь вершин и положим туда стартовую вершину.
    const planned = [s];
    color[s] = "gray";
    distance[s] = 0;
    while (planned.length > 0) {
        const u = planned.shift(); // Возьмём вершину из очереди.
        for (const v of adjList[u]) { // adjList - список смежности графа.
            if (color[v] === undefined) { // Серые и чёрные вершины уже
                // либо в очереди, либо обработаны.
                distance[v] = distance[u] + 1;
                previous[v] = u;
                color[v] = "gray";
                planned.push(v); // Запланируем посещение вершины.
            }
        }
        color[u] = "black"; // Теперь вершина считается обработанной.
    }
}

function ShortestPath(v) {
    // Класть вершины будем в стек, тогда
    // стартовая вершина окажется наверху стека
    // и порядок следования от s до v будет соответствовать
    // порядку извлечения вершин из стека.
    const path = [];
    let current_vertex = v;
    while (current_vertex !== null) { // Предшественник вершины s равен null.
        path.push(current_vertex);
        current_vertex = previous[current_vertex];
    }
    return path;
}
